asf_core_data/pipeline/mcs/process/mcs_epc_joining.py

```diff
@@ -0,0 +1,392 @@
+# File: heat_pump_adoption_modelling/pipeline/mcs/process/mcs_epc_joining.py
+"""Joining the MCS and EPC datasets.
+Overall process is as follows:
+- Standardise address and postcode fields
+- Extract numeric tokens from address
+- Group by postcode
+- Exact match on numeric tokens
+- Compare address using Jaro-Winkler score
+- Drop any match with score below a certain parameter
+- Of the remaining matches, take the ones with highest score
+- Join datasets using this matching
+"""
+
+import recordlinkage as rl
+
+from asf_core_data import PROJECT_DIR, get_yaml_config
+from asf_core_data.pipeline.mcs.process.process_mcs_installations import (
+    get_processed_installations_data,
+)
+from asf_core_data.getters.epc.epc_data import (
+    load_preprocessed_epc_data,
+)
+from asf_core_data.getters.data_getters import load_s3_data
+from asf_core_data.pipeline.mcs.process.process_mcs_utils import (
+    remove_punctuation,
+    extract_token_set,
+)
+
+config = get_yaml_config(PROJECT_DIR / "asf_core_data/config/base.yaml")
+
+max_token_length = config["MCS_EPC_MAX_TOKEN_LENGTH"]
+matching_parameter = config["MCS_EPC_MATCHING_PARAMETER"]
+bucket_name = config["BUCKET_NAME"]
+
+# ---------------------------------------------------------------------------------
+
+#### PREPROCESSING
+
+
+def prepare_hps(hps):
+    """Prepare Dataframe of HP installations by adding
+    standardised_postcode, standardised_address and numeric_tokens fields.
+
+    Args:
+        hps (Dataframe): Dataframe with postcode, address_1 and address_2 fields.
+
+    Returns:
+        Dataframe: Domestic HP records with added fields.
+    """
+
+    hps["standardised_postcode"] = (
+        hps["postcode"].fillna("unknown").str.upper().str.replace(" ", "")
+    )
+
+    hps["standardised_address"] = [
+        # Make address 1 and 2 lowercase, strip whitespace,
+        # and combine into a single string separated by a space
+        remove_punctuation(add1).lower().strip()
+        + " "
+        + remove_punctuation(add2).lower().strip()
+        for add1, add2 in zip(hps["address_1"].fillna(""), hps["address_2"].fillna(""))
```

Comment:
I think we did `fillna("")` for the address columns when loading, maybe not necessary anymore?

Reply:
I think that fillna was added to be cautious about not dropping NAs during deduplication. I might remove it from the loading script (and triple check that nothing gets dropped) and keep it here so that NAs are treated consistently in all fields

Reply:
Actually this is really ugly. I'll just remove them here